### PR TITLE
Merge devel to master, release 0.2.0

### PR DESCRIPTION
**Pull request by lrog**
Towards minor release 0.2.0 -- changes include:
* Added processing documents in bulk mode (endpoint: `/procees_bulk`)
* Added OpenAPI+Swagger specs
* Added tests
* Integration with TravisCI (builds + tests)
* Minor bugfixes

### CODE REVIEW

File: service/api-specs/openapi.json [hunk 1]
```diff
@@ -0,0 +1,293 @@
+{
+  "openapi": "3.0.1",
+  "info": {
+    "title": "NLP Service REST API",
+    "description": "Description of the initial NLP Service REST API. \n You can find out more the application at [http://github.com/CogStack/nlp-rest-service](http://github.com/CogStack/nlp-rest-service).",
```
Comment (afolarin) [inline service/api-specs/openapi.json]: `    "description": "Description of the initial NLP Service REST API. \n You can find out more about the application at [http://github.com/CogStack/nlp-rest-service](http://github.com/CogStack/nlp-rest-service).",
`

File: service/api-specs/openapi.json [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+            "description": "Success",
+            "content": {
+              "application/json": {
+                "schema": {
+                  "$ref": "#/components/schemas/ServiceResponseContent"
+                }
+              }
+            }
+          },
+          "400": {
```
Comment (afolarin) [inline service/api-specs/openapi.json:91]: question: how are you treating a partial fail in a bulk request?

File: service/api-specs/openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,185 @@
+openapi: 3.0.1
+info:
+  title: NLP Service REST API
+  description: "Description of the initial NLP Service REST API. \n You can find out\
```
Comment (afolarin) [inline service/api-specs/openapi.yaml]: just a style point, should we be consistent with the quoting in the YML file?

File: service/src/main/java/nlp/service/gate/service/GateNlpService.java [hunk 4]
```diff
@@ -95,19 +99,68 @@ public NlpProcessingResult process(NlpInputPayload payload, Map<String, String>
...
+            GateNlpResultDataMapper resultMapper = new GateNlpResultDataMapper();
+            resultMapper.setText(doc.getText());
+            resultMapper.setAnnotations(doc.getAnnotations());
+            resultMapper.setDocumentFeatures(doc.getDocumentFeatures());
+
+            // TODO: set in payload (if required):
+            // - binary document
+            // - enable/disable text inclusion
+
+            results.add(resultMapper.getProcessingResult());
```
Comment (afolarin) [inline service/src/main/java/nlp/service/gate/service/GateNlpService.java:154]: where is `getProcessingResult()`  set?

**Comment by afolarin** [comment]
also was wondering if bulk processing of n=1 would work for single document processing?

**Review by afolarin — APPROVED** [review]
LGTM.

File: service/src/main/java/nlp/service/gate/service/GateNlpService.java (hunk 4 above)
Comment (lrog) [reply service/src/main/java/nlp/service/gate/service/GateNlpService.java:154]: `GateNlpResultDataMapper` is a proxy class, serving as an interface for creating objects of `NlpProcessingResult` type (could have actually used a builder pattern here) that will be returned as the result of NLP processing according to the API specs. `getProcessingResult()` returns the initialised object of `NlpProcessingResult` type, where the fields were set above by calling `resultMapper.set*()`.
The same applies for `GateNlpContentDataMapper` and `NlpInputPayload`.

File: service/api-specs/openapi.json (hunk 2 above)
Comment (lrog) [reply service/api-specs/openapi.json:91]: In the current implementation of the Service Runner, when any of the documents fails, the whole batch is reported as failed with reported `400` return code.
It is actually an open question how shall this be addressed in a better and/or more informative way for the downstream app.

**Review by lrog — COMMENTED** [review]
> also was wondering if bulk processing of n=1 would work for single document processing?

Nope, 'n=1' case is not being examined for for single document processing (but a good point tho).

**Pull request merged**